### PR TITLE
chore(NODE-6160): sign and upload to releases

### PR DESCRIPTION
**Pull request by durran**
#### What is changing?
- Adds signing to the native bundle tarballs and uploads them to the Github release.
- Uploads the detached signatures for the tarballs to the Github release.
- Adds release please to the Github apps builds.
- Publishes the release with provenance.

##### Is there new documentation needed for these changes?

None

#### What is the motivation for this change?

NODE-6160

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [ ] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -61,21 +66,31 @@ jobs:
...
+      - name: actions/setup
+        uses: ./.github/actions/setup
+      - name: actions/sign_and_upload_package
+        uses: ./.github/actions/sign_and_upload_package
+        with: 
+          aws_role_arn: ${{ secrets.AWS_ROLE_ARN }}
+          aws_region_name: 'us-east-1'
+          aws_secret_id: ${{ secrets.AWS_SECRET_ID }}
+          npm_package_name: 'mongodb-client-encryption'
+      - run: npm publish --provenance
```
[the remark this answers is not in the document]
Comment (baileympearson) [reply .github/workflows/build.yml:94]: Signing should only depend on the contents of the tarball, it shouldn't make any difference.

[the remark this answers is not in the document]
Comment (nbbeeken) [reply .github/workflows/build.yml:94]: Nice, I thought there may be timestamp fields inside a tar that could change between the two archives, but if not, great!

**Review by nbbeeken — APPROVED** [review]
LGTM

File: .github/actions/sign_and_upload_package/action.yml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      run: for filename in build-*/*.tar.gz; do cp ${filename} artifacts/; done
+
+    - run: npm pack
+      shell: bash
+
+    - name: Get release version and release package file name
+      id: get_vars
+      shell: bash
+      run: |
+        package_version=$(jq --raw-output '.version' package.json)
```
Comment (aditi-khare-mongoDB) [inline .github/actions/sign_and_upload_package/action.yml:52]: code LGTM, is is possible to point me to an example test release, as an extra check?

Comment (durran) [reply .github/actions/sign_and_upload_package/action.yml:52]: What I'm going off is seeing release-please creating the PR now: https://github.com/mongodb-js/mongodb-client-encryption/pull/15 and was verifying the signing actions working but not being able to upload before release-please was put in, example run here: https://github.com/mongodb-js/mongodb-client-encryption/actions/runs/9403489046/job/25900735594

If that's not sufficient let me know and I can create a new fake repo and project and copy everything over.

Comment (aditi-khare-mongoDB) [reply .github/actions/sign_and_upload_package/action.yml:52]: No need for a fork: Can we have a run where we run `build.yml` to see the structure of the downloaded files, but comment out `gh release upload ..` and `npm publish`? To confirm the signatures are there.

**Review by aditi-khare-mongoDB — CHANGES_REQUESTED** [review]

**Comment by durran** [comment]
https://github.com/mongodb-js/mongodb-client-encryption/pull/9#discussion_r1638374731

@aditi-khare-mongoDB here:

**Comment by aditi-khare-mongoDB** [comment]
Thank! LGTM

**Review by aditi-khare-mongoDB — APPROVED** [review]

**Pull request merged**